### ISSUE TITLE
thanos sidecar is too restricted

### ISSUE THREAD

**Issue by uhthomas** [prometheus-operator/prometheus-operator#4057]
**What is missing?**

The Thanos sidecar section of the `Promethues` CRD is too restricted. It makes deployments which don't exactly align with what was envisioned almost impossible without implementing the sidecar manually.

**Why do we need it?**

Every organisation has different ways of operating applications, having the flexibility is important.

Take for example, I want to use [rook](https://rook.io/) and its object storage. The recommended way to mount the credentials and config is using `envFrom`. Consequently, the Thanos config would then look something [like this](https://github.com/uhthomas/automata/blob/f26292ae3d65d39fa6319cc74bf51d2340d71cd8/k8s/pillowtalk/thanos/compact/deployment_list.cue#L56-L65).

Using both `envFrom` and custom arguments are not supported, and so it's not possible to use this approach.

**Environment**

* Prometheus Operator version: 0.48.0

**Anything else we need to know?**:

**Comment by uhthomas** [comment]
So, upon reading the [PromethuesSpec](https://prometheus-operator.dev/docs/operator/api/#prometheusspec), specifically the `containers` attribute, it looks like some of this can be resolved.

This took a lot of digging, and doesn't quite solve the original problem but is a great step in the right direction.

**Comment by paulfantom** [comment]
Passing `objstore.config` to thanos sidecar is 100% possible without manipulating `containers` from PrometheusSpec. It is done with `objectStorageConfig`. Take a look at [`ThanosSpec`](https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#thanosspec) for full list of supported options of `PrometheusSpec.thanos` field.

**Comment by uhthomas** [comment]
It doesn't look like Thanos substitutes environment variables in its config which is what's making this hard.

I could be wrong, and even would very much like to be wrong 🙂 

The other option is to override the `args`, but that is not very nice.

**Comment by paulfantom** [comment]
> It doesn't look like Thanos substitutes environment variables in its config

No, it doesn't. Neither do prometheus. Hence I don't see why prometheus-operator should follow a different pattern. 

Additionally, the example you posted is from some deployment tool that injects those variables and creates a deployment with those variables built-in. This pattern is bad on many levels as some of those variables should be secret. Especially consider this line: https://github.com/uhthomas/automata/blob/f26292ae3d65d39fa6319cc74bf51d2340d71cd8/k8s/pillowtalk/thanos/compact/deployment_list.cue#L63. After such deployment is created you can easily get access key by doing simple `kubectl get deployment`.

**Comment by uhthomas** [comment]
That is not correct. The example I posed is completely transparent, the env substitutions (`$(BUCKET_NAME)`) are part of the actual `Deployment` 

```sh
cue export deployment_list.cue | jq -r '.deploymentList.items[].spec.template.spec.containers[].args'
[
  "compact",
  "-w",
  "--http-address=:80",
  "--http-grace-period=5s",
  "--objstore.config=type: S3\nconfig:\n  bucket: $(BUCKET_NAME)\n  endpoint: $(BUCKET_HOST):$(BUCKET_PORT)\n  region: $(BUCKET_REGION)\n  access_key: $(AWS_ACCESS_KEY_ID)\n  secret_key: $(AWS_SECRET_ACCESS_KEY)\n  insecure: true",
  "--downsampling.disable",
  "--deduplication.replica-label=replica",
  "--data-dir=/tmp/scratch"
]
```

Prometheus Operator creates a deployment containing `--objstore.config=$(OBJSTORE_CONFIG)` which means the env substitutions which would usually happen, don't

https://github.com/prometheus-operator/prometheus-operator/blob/53d01635ed08b7149e5265e9411c7e2489f188a4/pkg/thanos/statefulset.go#L256

**Comment by paulfantom** [comment]
Since in both cases one stores secrets in a Secret object, what would be the benefit of storing full config (and passing SecretKeySelector via `objectStorageConfig`) vs only doing a variable substitution? 

From maintainer pov, having `objectStorageConfig` is much simpler and creates simpler code hence it is preferable. Plus we don't need to update code whenever thanos changes config parameters. This means people deploying thanos also don't need to upgrade prometheus-operator when they wish to update thanos.

**Comment by uhthomas** [comment]
I understand. The problem in my case is the credentials for the object storage buckets are provisioned automatically by Rook. They are then available through a ConfigMap and Secret which are to be provided to containers as environment variables.

I don't really know what the solution is, but I will not be the only one to face this problem as Rook is, and will become, more commonplace.

**Comment by uhthomas** [comment]
It might be possible to use an init container to create `/aws-config` but that is once again, kind of hacky.

**Comment by paulfantom** [comment]
Seems like a possible solution would be to use sth like `ConfigMapSecret`. This way you can template a thanos config needed by prometheus-operator using values from secret provided by rook.

Example of such thing can be seen in https://github.com/thaum-xyz/ankhmorpork/blob/master/apps/monitoring/raw/alertmanager-secret.yaml (this is for alertmanager config).

**Comment by uhthomas** [comment]
Yeah, this is tricky.

Having to install an entire operator just to read environment variables for a Thanos config seems a bit much.

**Comment by paulfantom** [comment]
Yeah, I agree, this may be too much.

As far as I can see there are few more ways of fixing this and all of them have some issues:
1) prometheus-operator reads env variables and synthesizes thanos config from them - if thanos changes CLI parameters, then prometheus-operator needs an update to make it compatible.
2) thanos reads env variables directly - up to thanos maintainers. But generally picking up env variables isn't something that applications in prometheus ecosystem do. If implemented should be the simplest solution when it comes to implementations in ecosystem. @bwplotka wdyt?
3) rook allows templating output Secret - highly unlikely
4) init container picking up variables and creating a file - hacky
5) Better Secrets with built-in templating - good wish :D

**Comment by uhthomas** [comment]
FWIW I've opened an issue with Thanos asking what they think is the best way to use environment variables with configs.

https://github.com/thanos-io/thanos/issues/4293

**Comment by uhthomas** [comment]
Well, here's my hacky solution for the interim. It's using an init container to push the envsubst'd config into a shared emptyDir volume.

https://github.com/uhthomas/automata/blob/6bfe6628ccdc14915e2cd7242db9029d8422cbce/k8s/pillowtalk/prometheus/prometheus_list.cue

Thank you
